tensorcircuit/noisemodel.py

```diff
@@ -57,40 +59,79 @@ def add_noise(
+
+            def condition(d: Dict[str, Any]) -> bool:
+                return str(d["gatef"].n) == gate_name
+
+            if not isinstance(kraus, KrausList):
+                assert len(kraus) == 1
+                krauslist = kraus[0]
             else:
-                qubit_kraus["Default"] = kraus
+                krauslist = kraus
+            self.nc.append((description, condition, krauslist))
         else:
-            for i in range(len(qubit)):
-                if tuple(qubit[i]) in qubit_kraus:
-                    qubit_kraus[tuple(qubit[i])] = composedkraus(
-                        qubit_kraus[tuple(qubit[i])], kraus[i]
-                    )
-                else:
-                    if "Default" in qubit_kraus:
-                        qubit_kraus[tuple(qubit[i])] = composedkraus(
-                            qubit_kraus["Default"], kraus[i]
-                        )
-                    else:
-                        qubit_kraus[tuple(qubit[i])] = kraus[i]
+            for idx, krauslist in zip(qubit, kraus):
+                description = (gate_name, idx)
 
-        self.nc[gate_name] = qubit_kraus
+                # https://stackoverflow.com/questions/1107210/python-create-function-in-a-loop-capturing-the-loop-variable
+                def condition(  # type:ignore
+                    d: Dict[str, Any], _idx: Sequence[Any]
+                ) -> bool:
+                    # avoid bad black style because the long is too long
+                    b1 = d["gatef"].n == gate_name
+                    b2 = tuple(d["index"]) == tuple(_idx)
+                    return b1 and b2
 
-        if gate_name == "readout":
-            self.has_readout = True
-        else:
-            self.has_quantum = True
+                condition = partial(condition, _idx=idx)
+                self.nc.append((description, condition, krauslist))
+
+        self.has_quantum = True
+
+    def add_noise_by_condition(
+        self,
+        condition: Callable[[Dict[str, Any]], bool],
+        kraus: KrausList,
+        name: Optional[Any] = "custom",
+    ) -> None:
+        """
+        Add noise based on specified condition
+
+        :param condition: a function to decide if the noise should be added to the qir.
+        :type condition: Callable[[Dict[str, Any]], bool]
+        :param kraus: the error channel
+        :type kraus: KrausList
+        :param name: the name of the condition. A metadata that does not affect the numerics.
+        :type name: Any
+        """
+        self.nc.append((name, condition, kraus))
+
+    def channel_count(self, c: Circuit) -> int:
+        """
+        Count the total number of channels in a given circuit
+
+        :param c: the circuit to be counted
+        :type c: Circuit
+        :return: the count
+        :rtype: int
+        """
+        count = 0
+        for d in c.to_qir():
+            # pylint: disable-next=unused-variable
+            for _, condition, krauslist in self.nc:
```

Comment:
why we use krauslist instead of _ here?

Reply:
Didn't notice. I thought it's complaining about the first `_`. Will fix it in the next commit.